// couchbase/dist/binding.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CppTxnExternalException = exports.CppTxnFailureType = exports.CppViewSortOrder = exports.CppViewScanConsistency = exports.CppServiceType = exports.CppSearchScanConsistency = exports.CppSearchHighlightStyle = exports.CppQueryScanConsistency = exports.CppQueryProfileMode = exports.CppDiagPingState = exports.CppDiagEndpointState = exports.CppDiagClusterState = exports.CppDesignDocumentNamespace = exports.CppAnalyticsScanConsistency = exports.CppErrorNetworkErrc = exports.CppErrorFieldLevelEncryptionErrc = exports.CppErrorManagementErrc = exports.CppErrorViewErrc = exports.CppErrorSearchErrc = exports.CppErrorAnalyticsErrc = exports.CppErrorQueryErrc = exports.CppErrorKeyValueErrc = exports.CppErrorCommonErrc = exports.CppProtocolStatus = exports.CppProtocolDurabilityLevel = exports.CppProtocolMutateInRequestBodyStoreSemanticsType = exports.CppProtocolSubdocOpcode = exports.CppIoRetryReason = exports.CppManagementRbacAuthDomain = exports.CppManagementEventingFunctionProcessingStatus = exports.CppManagementEventingFunctionDeploymentStatus = exports.CppManagementEventingFunctionStatus = exports.CppManagementEventingFunctionBucketAccess = exports.CppManagementEventingFunctionLogLevel = exports.CppManagementEventingFunctionLanguageCompatibility = exports.CppManagementEventingFunctionDcpBoundary = exports.CppManagementClusterBucketStorageBackend = exports.CppManagementClusterBucketConflictResolution = exports.CppManagementClusterBucketEvictionPolicy = exports.CppManagementClusterBucketCompression = exports.CppManagementClusterBucketType = exports.CppManagementAnalyticsCouchbaseLinkEncryptionLevel = exports.zeroCas = void 0;
/*
eslint
jsdoc/require-jsdoc: off,
@typescript-eslint/no-empty-interface: off
*/
const bindings_1 = __importDefault(require("bindings"));
exports.zeroCas = 0;
//#region Autogenerated Bindings
var CppManagementAnalyticsCouchbaseLinkEncryptionLevel;
(function (CppManagementAnalyticsCouchbaseLinkEncryptionLevel) {
})(CppManagementAnalyticsCouchbaseLinkEncryptionLevel = exports.CppManagementAnalyticsCouchbaseLinkEncryptionLevel || (exports.CppManagementAnalyticsCouchbaseLinkEncryptionLevel = {}));
var CppManagementClusterBucketType;
(function (CppManagementClusterBucketType) {
})(CppManagementClusterBucketType = exports.CppManagementClusterBucketType || (exports.CppManagementClusterBucketType = {}));
var CppManagementClusterBucketCompression;
(function (CppManagementClusterBucketCompression) {
})(CppManagementClusterBucketCompression = exports.CppManagementClusterBucketCompression || (exports.CppManagementClusterBucketCompression = {}));
var CppManagementClusterBucketEvictionPolicy;
(function (CppManagementClusterBucketEvictionPolicy) {
})(CppManagementClusterBucketEvictionPolicy = exports.CppManagementClusterBucketEvictionPolicy || (exports.CppManagementClusterBucketEvictionPolicy = {}));
var CppManagementClusterBucketConflictResolution;
(function (CppManagementClusterBucketConflictResolution) {
})(CppManagementClusterBucketConflictResolution = exports.CppManagementClusterBucketConflictResolution || (exports.CppManagementClusterBucketConflictResolution = {}));
var CppManagementClusterBucketStorageBackend;
(function (CppManagementClusterBucketStorageBackend) {
})(CppManagementClusterBucketStorageBackend = exports.CppManagementClusterBucketStorageBackend || (exports.CppManagementClusterBucketStorageBackend = {}));
var CppManagementEventingFunctionDcpBoundary;
(function (CppManagementEventingFunctionDcpBoundary) {
})(CppManagementEventingFunctionDcpBoundary = exports.CppManagementEventingFunctionDcpBoundary || (exports.CppManagementEventingFunctionDcpBoundary = {}));
var CppManagementEventingFunctionLanguageCompatibility;
(function (CppManagementEventingFunctionLanguageCompatibility) {
})(CppManagementEventingFunctionLanguageCompatibility = exports.CppManagementEventingFunctionLanguageCompatibility || (exports.CppManagementEventingFunctionLanguageCompatibility = {}));
var CppManagementEventingFunctionLogLevel;
(function (CppManagementEventingFunctionLogLevel) {
})(CppManagementEventingFunctionLogLevel = exports.CppManagementEventingFunctionLogLevel || (exports.CppManagementEventingFunctionLogLevel = {}));
var CppManagementEventingFunctionBucketAccess;
(function (CppManagementEventingFunctionBucketAccess) {
})(CppManagementEventingFunctionBucketAccess = exports.CppManagementEventingFunctionBucketAccess || (exports.CppManagementEventingFunctionBucketAccess = {}));
var CppManagementEventingFunctionStatus;
(function (CppManagementEventingFunctionStatus) {
})(CppManagementEventingFunctionStatus = exports.CppManagementEventingFunctionStatus || (exports.CppManagementEventingFunctionStatus = {}));
var CppManagementEventingFunctionDeploymentStatus;
(function (CppManagementEventingFunctionDeploymentStatus) {
})(CppManagementEventingFunctionDeploymentStatus = exports.CppManagementEventingFunctionDeploymentStatus || (exports.CppManagementEventingFunctionDeploymentStatus = {}));
var CppManagementEventingFunctionProcessingStatus;
(function (CppManagementEventingFunctionProcessingStatus) {
})(CppManagementEventingFunctionProcessingStatus = exports.CppManagementEventingFunctionProcessingStatus || (exports.CppManagementEventingFunctionProcessingStatus = {}));
var CppManagementRbacAuthDomain;
(function (CppManagementRbacAuthDomain) {
})(CppManagementRbacAuthDomain = exports.CppManagementRbacAuthDomain || (exports.CppManagementRbacAuthDomain = {}));
var CppIoRetryReason;
(function (CppIoRetryReason) {
})(CppIoRetryReason = exports.CppIoRetryReason || (exports.CppIoRetryReason = {}));
var CppProtocolSubdocOpcode;
(function (CppProtocolSubdocOpcode) {
})(CppProtocolSubdocOpcode = exports.CppProtocolSubdocOpcode || (exports.CppProtocolSubdocOpcode = {}));
var CppProtocolMutateInRequestBodyStoreSemanticsType;
(function (CppProtocolMutateInRequestBodyStoreSemanticsType) {
})(CppProtocolMutateInRequestBodyStoreSemanticsType = exports.CppProtocolMutateInRequestBodyStoreSemanticsType || (exports.CppProtocolMutateInRequestBodyStoreSemanticsType = {}));
var CppProtocolDurabilityLevel;
(function (CppProtocolDurabilityLevel) {
})(CppProtocolDurabilityLevel = exports.CppProtocolDurabilityLevel || (exports.CppProtocolDurabilityLevel = {}));
var CppProtocolStatus;
(function (CppProtocolStatus) {
})(CppProtocolStatus = exports.CppProtocolStatus || (exports.CppProtocolStatus = {}));
var CppErrorCommonErrc;
(function (CppErrorCommonErrc) {
})(CppErrorCommonErrc = exports.CppErrorCommonErrc || (exports.CppErrorCommonErrc = {}));
var CppErrorKeyValueErrc;
(function (CppErrorKeyValueErrc) {
})(CppErrorKeyValueErrc = exports.CppErrorKeyValueErrc || (exports.CppErrorKeyValueErrc = {}));
var CppErrorQueryErrc;
(function (CppErrorQueryErrc) {
})(CppErrorQueryErrc = exports.CppErrorQueryErrc || (exports.CppErrorQueryErrc = {}));
var CppErrorAnalyticsErrc;
(function (CppErrorAnalyticsErrc) {
})(CppErrorAnalyticsErrc = exports.CppErrorAnalyticsErrc || (exports.CppErrorAnalyticsErrc = {}));
var CppErrorSearchErrc;
(function (CppErrorSearchErrc) {
})(CppErrorSearchErrc = exports.CppErrorSearchErrc || (exports.CppErrorSearchErrc = {}));
var CppErrorViewErrc;
(function (CppErrorViewErrc) {
})(CppErrorViewErrc = exports.CppErrorViewErrc || (exports.CppErrorViewErrc = {}));
var CppErrorManagementErrc;
(function (CppErrorManagementErrc) {
})(CppErrorManagementErrc = exports.CppErrorManagementErrc || (exports.CppErrorManagementErrc = {}));
var CppErrorFieldLevelEncryptionErrc;
(function (CppErrorFieldLevelEncryptionErrc) {
})(CppErrorFieldLevelEncryptionErrc = exports.CppErrorFieldLevelEncryptionErrc || (exports.CppErrorFieldLevelEncryptionErrc = {}));
var CppErrorNetworkErrc;
(function (CppErrorNetworkErrc) {
})(CppErrorNetworkErrc = exports.CppErrorNetworkErrc || (exports.CppErrorNetworkErrc = {}));
var CppAnalyticsScanConsistency;
(function (CppAnalyticsScanConsistency) {
})(CppAnalyticsScanConsistency = exports.CppAnalyticsScanConsistency || (exports.CppAnalyticsScanConsistency = {}));
var CppDesignDocumentNamespace;
(function (CppDesignDocumentNamespace) {
})(CppDesignDocumentNamespace = exports.CppDesignDocumentNamespace || (exports.CppDesignDocumentNamespace = {}));
var CppDiagClusterState;
(function (CppDiagClusterState) {
})(CppDiagClusterState = exports.CppDiagClusterState || (exports.CppDiagClusterState = {}));
var CppDiagEndpointState;
(function (CppDiagEndpointState) {
})(CppDiagEndpointState = exports.CppDiagEndpointState || (exports.CppDiagEndpointState = {}));
var CppDiagPingState;
(function (CppDiagPingState) {
})(CppDiagPingState = exports.CppDiagPingState || (exports.CppDiagPingState = {}));
var CppQueryProfileMode;
(function (CppQueryProfileMode) {
})(CppQueryProfileMode = exports.CppQueryProfileMode || (exports.CppQueryProfileMode = {}));
var CppQueryScanConsistency;
(function (CppQueryScanConsistency) {
})(CppQueryScanConsistency = exports.CppQueryScanConsistency || (exports.CppQueryScanConsistency = {}));
var CppSearchHighlightStyle;
(function (CppSearchHighlightStyle) {
})(CppSearchHighlightStyle = exports.CppSearchHighlightStyle || (exports.CppSearchHighlightStyle = {}));
var CppSearchScanConsistency;
(function (CppSearchScanConsistency) {
})(CppSearchScanConsistency = exports.CppSearchScanConsistency || (exports.CppSearchScanConsistency = {}));
var CppServiceType;
(function (CppServiceType) {
})(CppServiceType = exports.CppServiceType || (exports.CppServiceType = {}));
var CppViewScanConsistency;
(function (CppViewScanConsistency) {
})(CppViewScanConsistency = exports.CppViewScanConsistency || (exports.CppViewScanConsistency = {}));
var CppViewSortOrder;
(function (CppViewSortOrder) {
})(CppViewSortOrder = exports.CppViewSortOrder || (exports.CppViewSortOrder = {}));
var CppTxnFailureType;
(function (CppTxnFailureType) {
})(CppTxnFailureType = exports.CppTxnFailureType || (exports.CppTxnFailureType = {}));
var CppTxnExternalException;
(function (CppTxnExternalException) {
})(CppTxnExternalException = exports.CppTxnExternalException || (exports.CppTxnExternalException = {}));
// Load it with require
const binding = (0, bindings_1.default)('couchbase_impl');
exports.default = binding;
