// couchbase/dist/collection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collection = void 0;
const binarycollection_1 = require("./binarycollection");
const binding_1 = __importStar(require("./binding"));
const bindingutilities_1 = require("./bindingutilities");
const crudoptypes_1 = require("./crudoptypes");
const datastructures_1 = require("./datastructures");
const errors_1 = require("./errors");
const generaltypes_1 = require("./generaltypes");
const sdspecs_1 = require("./sdspecs");
const sdutils_1 = require("./sdutils");
const utilities_1 = require("./utilities");
/**
 * Exposes the operations which are available to be performed against a collection.
 * Namely the ability to perform KV operations.
 *
 * @category Core
 */
class Collection {
    /**
    @internal
    */
    constructor(scope, collectionName) {
        this._scope = scope;
        this._name = collectionName;
        this._conn = scope.conn;
    }
    /**
     * @internal
     */
    static get DEFAULT_NAME() {
        return '_default';
    }
    /**
    @internal
    */
    get conn() {
        return this._conn;
    }
    /**
    @internal
    */
    get cluster() {
        return this._scope.bucket.cluster;
    }
    /**
    @internal
    */
    get scope() {
        return this._scope;
    }
    /**
    @internal
    */
    get transcoder() {
        return this._scope.transcoder;
    }
    /**
    @internal
    */
    _mutationTimeout(durabilityLevel) {
        if (durabilityLevel !== undefined &&
            durabilityLevel !== null &&
            durabilityLevel !== generaltypes_1.DurabilityLevel.None) {
            return this.cluster.kvDurableTimeout;
        }
        return this.cluster.kvTimeout;
    }
    /**
     * @internal
     */
    _cppDocId(key) {
        return {
            bucket: this.scope.bucket.name,
            scope: this.scope.name || '_default',
            collection: this.name || '_default',
            key: key,
        };
    }
    /**
     * @internal
     */
    _encodeDoc(transcoder, value, callback) {
        try {
            // BUG(JSCBC-1054): We should avoid doing buffer conversion.
            const [bytesBuf, flagsOut] = transcoder.encode(value);
            const bytesStr = bytesBuf.toString('binary');
            callback(null, bytesStr, flagsOut);
        }
        catch (e) {
            return callback(e, '', 0);
        }
    }
    /**
     * @internal
     */
    _decodeDoc(transcoder, bytesStr, flags, callback) {
        try {
            // BUG(JSCBC-1054): We should avoid doing buffer conversion.
            const bytes = Buffer.from(bytesStr, 'binary');
            const content = transcoder.decode(bytes, flags);
            callback(null, content);
        }
        catch (e) {
            return callback(e, null);
        }
    }
    /**
     * The name of the collection this Collection object references.
     */
    get name() {
        return this._name;
    }
    /**
     * Retrieves the value of a document from the collection.
     *
     * @param key The document key to retrieve.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    get(key, options, callback) {
        if (options instanceof Function) {
            callback = arguments[1];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        if (options.project || options.withExpiry) {
            return this._projectedGet(key, options, callback);
        }
        const transcoder = options.transcoder || this.transcoder;
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.get({
                id: this._cppDocId(key),
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                this._decodeDoc(transcoder, resp.value, resp.flags, (err, content) => {
                    if (err) {
                        return wrapCallback(err, null);
                    }
                    wrapCallback(null, new crudoptypes_1.GetResult({
                        content: content,
                        cas: resp.cas,
                    }));
                });
            });
        }, callback);
    }
    _projectedGet(key, options, callback) {
        let expiryStart = -1;
        let projStart = -1;
        let paths = [];
        let spec = [];
        let needReproject = false;
        if (options.withExpiry) {
            expiryStart = spec.length;
            spec.push(sdspecs_1.LookupInSpec.get(sdspecs_1.LookupInMacro.Expiry));
        }
        projStart = spec.length;
        if (!options.project) {
            paths = [''];
            spec.push(sdspecs_1.LookupInSpec.get(''));
        }
        else {
            let projects = options.project;
            if (!Array.isArray(projects)) {
                projects = [projects];
            }
            for (let i = 0; i < projects.length; ++i) {
                paths.push(projects[i]);
                spec.push(sdspecs_1.LookupInSpec.get(projects[i]));
            }
        }
        // The following code relies on the projections being
        // the last segment of the specs array, this way we handle
        // an overburdened operation in a single area.
        if (spec.length > 16) {
            spec = spec.splice(0, projStart);
            spec.push(sdspecs_1.LookupInSpec.get(''));
            needReproject = true;
        }
        return utilities_1.PromiseHelper.wrapAsync(async () => {
            const res = await this.lookupIn(key, spec, {
                ...options,
            });
            let content = null;
            let expiry = undefined;
            if (expiryStart >= 0) {
                const expiryRes = res.content[expiryStart];
                expiry = expiryRes.value;
            }
            if (projStart >= 0) {
                if (!needReproject) {
                    for (let i = 0; i < paths.length; ++i) {
                        const projPath = paths[i];
                        const projRes = res.content[projStart + i];
                        if (!projRes.error) {
                            content = sdutils_1.SdUtils.insertByPath(content, projPath, projRes.value);
                        }
                    }
                }
                else {
                    content = {};
                    const reprojRes = res.content[projStart];
                    for (let j = 0; j < paths.length; ++j) {
                        const reprojPath = paths[j];
                        const value = sdutils_1.SdUtils.getByPath(reprojRes.value, reprojPath);
                        content = sdutils_1.SdUtils.insertByPath(content, reprojPath, value);
                    }
                }
            }
            return new crudoptypes_1.GetResult({
                content: content,
                cas: res.cas,
                expiryTime: expiry,
            });
        }, callback);
    }
    /**
     * Checks whether a specific document exists or not.
     *
     * @param key The document key to check for existence.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    exists(key, options, callback) {
        if (options instanceof Function) {
            callback = arguments[1];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.exists({
                id: this._cppDocId(key),
                partition: 0,
                opaque: 0,
                timeout,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                // BUG(JSCBC-1006): Remove this workaround when the underlying bug is resolved.
                if (err instanceof errors_1.DocumentNotFoundError) {
                    return wrapCallback(null, new crudoptypes_1.ExistsResult({
                        cas: undefined,
                        exists: false,
                    }));
                }
                if (err) {
                    return wrapCallback(err, null);
                }
                if (resp.deleted) {
                    return wrapCallback(null, new crudoptypes_1.ExistsResult({
                        cas: undefined,
                        exists: false,
                    }));
                }
                wrapCallback(null, new crudoptypes_1.ExistsResult({
                    cas: resp.cas,
                    exists: true,
                }));
            });
        }, callback);
    }
    /**
     * Inserts a new document to the collection, failing if the document already exists.
     *
     * @param key The document key to insert.
     * @param value The value of the document to insert.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    insert(key, value, options, callback) {
        if (options instanceof Function) {
            callback = arguments[3];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const expiry = options.expiry;
        const transcoder = options.transcoder || this.transcoder;
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this._mutationTimeout(durabilityLevel);
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._encodeDoc(transcoder, value, (err, bytes, flags) => {
                if (err) {
                    return wrapCallback(err, null);
                }
                this._conn.insert({
                    id: this._cppDocId(key),
                    value: bytes,
                    flags,
                    expiry: expiry || 0,
                    durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                    timeout,
                    partition: 0,
                    opaque: 0,
                }, (cppErr, resp) => {
                    const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                    if (err) {
                        return wrapCallback(err, null);
                    }
                    wrapCallback(err, new crudoptypes_1.MutationResult({
                        cas: resp.cas,
                        token: resp.token,
                    }));
                });
            });
        }, callback);
    }
    /**
     * Upserts a document to the collection.  This operation succeeds whether or not the
     * document already exists.
     *
     * @param key The document key to upsert.
     * @param value The new value for the document.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    upsert(key, value, options, callback) {
        if (options instanceof Function) {
            callback = arguments[3];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const expiry = options.expiry;
        const preserve_expiry = options.preserveExpiry;
        const transcoder = options.transcoder || this.transcoder;
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this._mutationTimeout(durabilityLevel);
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            let bytes, flags;
            try {
                // BUG(JSCBC-1054): We should avoid doing buffer conversion.
                const [bytesStr, flagsOut] = transcoder.encode(value);
                flags = flagsOut;
                bytes = bytesStr.toString('binary');
            }
            catch (e) {
                return wrapCallback(e, null);
            }
            this._conn.upsert({
                id: this._cppDocId(key),
                value: bytes,
                flags,
                expiry: expiry || 0,
                preserve_expiry: preserve_expiry || false,
                durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                wrapCallback(err, new crudoptypes_1.MutationResult({
                    cas: resp.cas,
                    token: resp.token,
                }));
            });
        }, callback);
    }
    /**
     * Replaces the value of an existing document.  Failing if the document does not exist.
     *
     * @param key The document key to replace.
     * @param value The new value for the document.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    replace(key, value, options, callback) {
        if (options instanceof Function) {
            callback = arguments[3];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const expiry = options.expiry || 0;
        const cas = options.cas;
        const preserve_expiry = options.preserveExpiry;
        const transcoder = options.transcoder || this.transcoder;
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this._mutationTimeout(durabilityLevel);
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._encodeDoc(transcoder, value, (err, bytes, flags) => {
                if (err) {
                    return wrapCallback(err, null);
                }
                this._conn.replace({
                    id: this._cppDocId(key),
                    value: bytes,
                    flags,
                    expiry,
                    cas: cas || binding_1.zeroCas,
                    preserve_expiry: preserve_expiry || false,
                    durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                    timeout,
                    partition: 0,
                    opaque: 0,
                }, (cppErr, resp) => {
                    const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                    if (err) {
                        return wrapCallback(err, null);
                    }
                    wrapCallback(err, new crudoptypes_1.MutationResult({
                        cas: resp.cas,
                        token: resp.token,
                    }));
                });
            });
        }, callback);
    }
    /**
     * Remove an existing document from the collection.
     *
     * @param key The document key to remove.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    remove(key, options, callback) {
        if (options instanceof Function) {
            callback = arguments[1];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const cas = options.cas;
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this._mutationTimeout(durabilityLevel);
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.remove({
                id: this._cppDocId(key),
                cas: cas || binding_1.zeroCas,
                durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                wrapCallback(err, new crudoptypes_1.MutationResult({
                    cas: resp.cas,
                }));
            });
        }, callback);
    }
    /**
     * Retrieves the value of the document and simultanously updates the expiry time
     * for the same document.
     *
     * @param key The document to fetch and touch.
     * @param expiry The new expiry to apply to the document, specified in seconds.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    getAndTouch(key, expiry, options, callback) {
        if (options instanceof Function) {
            callback = arguments[2];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const transcoder = options.transcoder || this.transcoder;
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.getAndTouch({
                id: this._cppDocId(key),
                expiry,
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                this._decodeDoc(transcoder, resp.value, resp.flags, (err, content) => {
                    if (err) {
                        return wrapCallback(err, null);
                    }
                    wrapCallback(err, new crudoptypes_1.GetResult({
                        content: content,
                        cas: resp.cas,
                    }));
                });
            });
        }, callback);
    }
    /**
     * Updates the expiry on an existing document.
     *
     * @param key The document key to touch.
     * @param expiry The new expiry to set for the document, specified in seconds.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    touch(key, expiry, options, callback) {
        if (options instanceof Function) {
            callback = arguments[2];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.touch({
                id: this._cppDocId(key),
                expiry,
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                wrapCallback(err, new crudoptypes_1.MutationResult({
                    cas: resp.cas,
                }));
            });
        }, callback);
    }
    /**
     * Locks a document and retrieves the value of that document at the time it is locked.
     *
     * @param key The document key to retrieve and lock.
     * @param lockTime The amount of time to lock the document for, specified in seconds.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    getAndLock(key, lockTime, options, callback) {
        if (options instanceof Function) {
            callback = arguments[2];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const transcoder = options.transcoder || this.transcoder;
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.getAndLock({
                id: this._cppDocId(key),
                lock_time: lockTime,
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                this._decodeDoc(transcoder, resp.value, resp.flags, (err, content) => {
                    if (err) {
                        return wrapCallback(err, null);
                    }
                    wrapCallback(err, new crudoptypes_1.GetResult({
                        cas: resp.cas,
                        content: content,
                    }));
                });
            });
        }, callback);
    }
    /**
     * Unlocks a previously locked document.
     *
     * @param key The document key to unlock.
     * @param cas The CAS of the document, used to validate lock ownership.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    unlock(key, cas, options, callback) {
        if (options instanceof Function) {
            callback = arguments[2];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.unlock({
                id: this._cppDocId(key),
                cas,
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err);
                }
                wrapCallback(null);
            });
        }, callback);
    }
    /**
     * Performs a lookup-in operation against a document, fetching individual fields or
     * information about specific fields inside the document value.
     *
     * @param key The document key to look in.
     * @param specs A list of specs describing the data to fetch from the document.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    lookupIn(key, specs, options, callback) {
        if (options instanceof Function) {
            callback = arguments[2];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const cppSpecs = [];
        for (let i = 0; i < specs.length; ++i) {
            cppSpecs.push({
                opcode: specs[i]._op,
                flags: specs[i]._flags,
                path: specs[i]._path,
                original_index: i,
            });
        }
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.lookupIn({
                id: this._cppDocId(key),
                specs: {
                    entries: cppSpecs,
                },
                timeout,
                partition: 0,
                opaque: 0,
                access_deleted: false,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (resp && resp.fields) {
                    const content = [];
                    for (let i = 0; i < resp.fields.length; ++i) {
                        const itemRes = resp.fields[i];
                        let error = (0, bindingutilities_1.errorFromCpp)(itemRes.ec);
                        let value = undefined;
                        if (itemRes.value && itemRes.value.length > 0) {
                            value = JSON.parse(itemRes.value);
                        }
                        // BUG(JSCBC-1016): Should remove this workaround when the underlying bug is resolved.
                        if (itemRes.opcode === binding_1.default.protocol_subdoc_opcode.exists) {
                            error = null;
                            value = itemRes.exists;
                        }
                        content.push(new crudoptypes_1.LookupInResultEntry({
                            error,
                            value,
                        }));
                    }
                    wrapCallback(err, new crudoptypes_1.LookupInResult({
                        content: content,
                        cas: resp.cas,
                    }));
                    return;
                }
                wrapCallback(err, null);
            });
        }, callback);
    }
    /**
     * Performs a mutate-in operation against a document.  Allowing atomic modification of
     * specific fields within a document.  Also enables access to document extended-attributes.
     *
     * @param key The document key to mutate.
     * @param specs A list of specs describing the operations to perform on the document.
     * @param options Optional parameters for this operation.
     * @param callback A node-style callback to be invoked after execution.
     */
    mutateIn(key, specs, options, callback) {
        if (options instanceof Function) {
            callback = arguments[2];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const cppSpecs = [];
        for (let i = 0; i < specs.length; ++i) {
            cppSpecs.push({
                opcode: specs[i]._op,
                flags: specs[i]._flags,
                path: specs[i]._path,
                param: specs[i]._data,
                original_index: 0,
            });
        }
        const storeSemantics = options.upsertDocument
            ? generaltypes_1.StoreSemantics.Upsert
            : options.storeSemantics;
        const expiry = options.expiry;
        const preserveExpiry = options.preserveExpiry;
        const cas = options.cas;
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this._mutationTimeout(durabilityLevel);
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.mutateIn({
                id: this._cppDocId(key),
                store_semantics: (0, bindingutilities_1.storeSemanticToCpp)(storeSemantics),
                specs: {
                    entries: cppSpecs,
                },
                expiry,
                preserve_expiry: preserveExpiry || false,
                cas: cas || binding_1.zeroCas,
                durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                timeout,
                partition: 0,
                opaque: 0,
                access_deleted: false,
                create_as_deleted: false,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (resp && resp.fields) {
                    const content = [];
                    for (let i = 0; i < resp.fields.length; ++i) {
                        const itemRes = resp.fields[i];
                        let value = undefined;
                        if (itemRes.value && itemRes.value.length > 0) {
                            value = JSON.parse(itemRes.value);
                        }
                        content.push(new crudoptypes_1.MutateInResultEntry({
                            value,
                        }));
                    }
                    wrapCallback(err, new crudoptypes_1.MutateInResult({
                        content: content,
                        cas: resp.cas,
                    }));
                    return;
                }
                wrapCallback(err, null);
            });
        }, callback);
    }
    /**
     * Returns a CouchbaseList permitting simple list storage in a document.
     *
     * @param key The document key the data-structure resides in.
     */
    list(key) {
        return new datastructures_1.CouchbaseList(this, key);
    }
    /**
     * Returns a CouchbaseQueue permitting simple queue storage in a document.
     *
     * @param key The document key the data-structure resides in.
     */
    queue(key) {
        return new datastructures_1.CouchbaseQueue(this, key);
    }
    /**
     * Returns a CouchbaseMap permitting simple map storage in a document.
     *
     * @param key The document key the data-structure resides in.
     */
    map(key) {
        return new datastructures_1.CouchbaseMap(this, key);
    }
    /**
     * Returns a CouchbaseSet permitting simple set storage in a document.
     *
     * @param key The document key the data-structure resides in.
     */
    set(key) {
        return new datastructures_1.CouchbaseSet(this, key);
    }
    /**
     * Returns a BinaryCollection object reference, allowing access to various
     * binary operations possible against a collection.
     */
    binary() {
        return new binarycollection_1.BinaryCollection(this);
    }
    /**
     * @internal
     */
    _binaryIncrement(key, delta, options, callback) {
        if (options instanceof Function) {
            callback = arguments[2];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const initial_value = options.initial;
        const expiry = options.expiry;
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.increment({
                id: this._cppDocId(key),
                delta,
                initial_value,
                expiry: expiry || 0,
                durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                wrapCallback(err, new crudoptypes_1.CounterResult({
                    cas: resp.cas,
                    token: resp.token,
                    value: resp.content,
                }));
            });
        }, callback);
    }
    /**
     * @internal
     */
    _binaryDecrement(key, delta, options, callback) {
        if (options instanceof Function) {
            callback = arguments[2];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const initial_value = options.initial;
        const expiry = options.expiry;
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            this._conn.decrement({
                id: this._cppDocId(key),
                delta,
                initial_value,
                expiry: expiry || 0,
                durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                wrapCallback(err, new crudoptypes_1.CounterResult({
                    cas: resp.cas,
                    token: resp.token,
                    value: resp.content,
                }));
            });
        }, callback);
    }
    /**
     * @internal
     */
    _binaryAppend(key, value, options, callback) {
        if (options instanceof Function) {
            callback = arguments[3];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            // BUG(JSCBC-1054): We should avoid doing buffer conversion.
            const bytesStr = value.toString('binary');
            this._conn.append({
                id: this._cppDocId(key),
                value: bytesStr,
                durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                wrapCallback(err, new crudoptypes_1.MutationResult({
                    cas: resp.cas,
                    token: resp.token,
                }));
            });
        }, callback);
    }
    /**
     * @internal
     */
    _binaryPrepend(key, value, options, callback) {
        if (options instanceof Function) {
            callback = arguments[3];
            options = undefined;
        }
        if (!options) {
            options = {};
        }
        const durabilityLevel = options.durabilityLevel;
        const timeout = options.timeout || this.cluster.kvTimeout;
        return utilities_1.PromiseHelper.wrap((wrapCallback) => {
            // BUG(JSCBC-1054): We should avoid doing buffer conversion.
            const bytesStr = value.toString('binary');
            this._conn.prepend({
                id: this._cppDocId(key),
                value: bytesStr,
                durability_level: (0, bindingutilities_1.durabilityToCpp)(durabilityLevel),
                timeout,
                partition: 0,
                opaque: 0,
            }, (cppErr, resp) => {
                const err = (0, bindingutilities_1.errorFromCpp)(cppErr);
                if (err) {
                    return wrapCallback(err, null);
                }
                wrapCallback(err, new crudoptypes_1.MutationResult({
                    cas: resp.cas,
                    token: resp.token,
                }));
            });
        }, callback);
    }
}
exports.Collection = Collection;
